src/main/java/mate/academy/Dao/BookDaoImpl.java

```diff
@@ -0,0 +1,117 @@
+            }
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                Long id = resultSet.getObject(INDEX_ID, Long.class);
+                book.setId(id);
+            }
+        } catch (SQLException e) {
+            throw new DataException("Can't create book = " + book, e);
+        }
+        return book;
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection();
+             PreparedStatement statement = connection.prepareStatement(SELECT_BY_ID)) {
+            statement.setLong(INDEX_ID, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(getBook(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataException("Can't find book with id = " + id, e);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try (Connection connection = ConnectionUtil.getConnection();
+             PreparedStatement statement = connection.prepareStatement(SELECT_ALL)) {
+            ResultSet resultSet = statement.executeQuery();
+            List<Book> result = new ArrayList<>();
+            while (resultSet.next()) {
+                result.add(getBook(resultSet));
+            }
+            return result;
+        } catch (SQLException e) {
+            throw new DataException("Can't retrieve all books", e);
+        }
+    }
+
+    @Override
+    public Book update(Book book) {
+        try (Connection connection = ConnectionUtil.getConnection();
+             PreparedStatement statement = connection.prepareStatement(UPDATE)) {
+            statement.setString(TITLE_PART, book.getTitle());
+            statement.setObject(PRICE_PART, book.getPrice());
+            statement.setLong(ID_PART, book.getId());
+            if (statement.executeUpdate() < MINIMUM_CHANGES) {
+                throw new RuntimeException("Can't insert less than one row");
+            }
+        } catch (SQLException e) {
+            throw new DataException("Can't update the book = " + book, e);
+        }
+        return book;
+    }
+
+    @Override
+    public boolean deleteById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection();
+             PreparedStatement statement = connection.prepareStatement(DELETE_BY_ID)) {
+            statement.setLong(INDEX_ID, id);
+            return statement.executeUpdate() > 0;
+        } catch (SQLException e) {
+            throw new DataException("Can't delete book with id = " + id, e);
+        }
+    }
+
+    private Book getBook(ResultSet resultSet) {
+        Book book = new Book();
+        try {
+            book.setId(resultSet.getObject("id", Long.class));
+            book.setTitle(resultSet.getString("title"));
+            book.setPrice(resultSet.getBigDecimal("price") != null ?
```

Comment:
Is it better to return null if there is not price, because it explicitly says thas price in not defined, price 0 means that price exists and it is 0

Reply:
not resolved